### ISSUE TITLE
Function signature not being formatted properly

### ISSUE THREAD

**Issue by jonhoo** [rust-lang-nursery/rustfmt#104]
Currently, a line like:

```rust
pub fn fset(bytes : Vec<u8>, flags : u32, expires : i64, casid : u64) -> Box<Fn(Option<&Item>) -> (Status, Result<Item, Option<String>>)> {
```

causes rustfmt to choke, saying "Rustfmt couldn't fix (sorry). abc.rs:29: line longer than 100 characters". However, the Rust style guide has a [section on line wrapping](https://aturon.github.io/style/whitespace.html#line-wrapping) that could probably be automated.

**Comment by jonhoo** [comment]
FWIW, it seems like we already do this, just not for long lines?
Also, the whitespace handling for function arguments seems a bit off?

```diff
-    fn op(&self, key : &[u8], upd : Box<Fn(Option<&memcache::Item>) -> (memcache::Status, Result<memcache::Item, Option<String>>)>) -> MapResult {
+    fn op(&self,
+          key: &[u8],
+          upd: Box<Fn(Option<&memcache::Item>) ->
+    (memcache::Status, Result<memcache::Item, Option<String>>)>)
+          -> MapResult {
```

**Comment by nrc** [comment]
One of the major goals of rustfmt is to end up with lines of the right length. However, for any given line, how to actually split the line is a very difficult question which depends on what the line is. As in the example you gave, we already do it for function declarations, and some other items, Eventually, we'll do it for everything. In the meantime we give a warning when we can't (yet) fix something.

The whitespace error you notice there looks like it is due to a function type in a function signature and I think is a bug. It would be great if you could create a new issue for that specific problem.

Closing this issue because we want to fix it, but the issue is not specific enough to be actionable.

**Comment by jonhoo** [comment]
@nrc: The line I give in the original post **is** a function declaration (and so should be wrappable), but it is not being wrapped.

**Comment by nrc** [comment]
Ah, sorry, I misunderstood the report. I reopened the issue and edited the title to try and make it a bit more specific. Thanks for the bug report!

**Comment by jonhoo** [comment]
No worries -- it was poorly phrased as I didn't know that the line-wrapping was supposed to work for function definitions until my second post.

**Comment by cassiersg** [comment]
I can't reproduce your bug : for a file whose content is only the line you provided :
```rust
pub fn fset(bytes : Vec<u8>, flags : u32, expires : i64, casid : u64) -> Box<Fn(Option<&Item>) -> (Status, Result<Item, Option<String>>)> {}
```
rustfmt gives
```rust

pub fn fset(bytes: Vec<u8>,
            flags: u32,
            expires: i64,
            casid: u64)
            -> Box<Fn(Option<&Item>) -> (Status, Result<Item, Option<String>>)> {
}
```
and no warning.

Can you give an example where rustfmt fails ?

**Comment by marcusklaas** [comment]
The problem occurs at deeper indentation levels. The return type is never split over multiple lines, exceeding the column limit.

**Comment by jonhoo** [comment]
Yup. Specifically, try:

```rust
impl Map {
    fn op(&self, key : &[u8], upd : Box<Fn(Option<&memcache::Item>) -> (memcache::Status, Result<memcache::Item, Option<String>>)>) -> MapResult {
    }
}
```

**Comment by cassiersg** [comment]
This doesn't work because we don't yet fully reformat types. (#180)

**Comment by jonhoo** [comment]
Presumably now related to #350.

**Comment by nrc** [comment]
Fixed by #637
